BeepBoopHQ/starter-python-bot
# Bot Rewrite and Multi team bot support

#### What's this PR do?

Complete rewrite of the python slack bot (doing away with the confusing plugins architecture, and decomposing responsibilities into classes) to hopefully make it simpler and more approachable.  Also, providing the expected interface for the Python beepboop package, so that it's BotManager and delegate bot creation when the Resourcer adds/updates/removes teams.
#### Where should the reviewer start?

`app.py` - entry point, which has two paths (if SLACK_TOKEN in env, then single instance dev mode, else depend on beepboop resourcer and bot_manager to be responsible for instantiating bot instances)
`bot/slack_bot.py` - Instantiates the necessary Slack clients (RTM and Web) after receiving an access token, and then polls RTM for messages to hand off to the `RtmEventHandler`.
`bot/slack_clients.py` - API facade to `slacker`(web api) and `slackclient`(rtm api)
`bot/event_handler.py` - Event matcher and delegator for messaging
`bot/messenger.py` - Responsible for constructing and sending messages back to slack on behalf of the bot.
#### How should this be manually tested?

`virtualenv venv`
`source venv/bin/activate`
`pip install -r requirements.txt`

Multi team mode:
Sign up on: https://preview.beepboophq.com
Setup a new project (doesn't need to point to this repository, just need to get tokens and access Dev mode tab to add new teams)
Get and apply tokens (e.g. `export BEEPBOOP_ID=7b44256c235d4737a43d21f5ad456b09 BEEPBOOP_RESOURCER=wss://beepbooppreview.rnp.io/ws BEEPBOOP_TOKEN=eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJQSUQiOiI3YjQ0MjU2YzIzNWQ0NzM3YTQzZDIxZjVhZDQ1NmIwOSIsIlMiOjQwNjQ2OTg0MX0.xJtow8RPOYdh1mH0OKqxmHmeF-BzQZcCMAp47XrqRmezD2IxwpVyqVJdQYapFauO8Bam6K17iczzg1o1D-AGsA`)
Run the bot: `./bot/app.py`
Add a team and bot token.
Verify in console logs bot is instantiated after receiving add_resource message, and then you can chat with it: `hi @<py-bot-name>`
Add another team and bot token.
Verify that each bot handles messaging independently.

Local one instance dev mode:
Export slack token and run bot:
`export SLACK_TOKEN=<bot_token>; ./bot/app.py`
